### PR TITLE
Document parallelism and thread scheduling in the architecture guide

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of #7013 

## Rationale for this change

The topic of thread scheduling and why not use Rayon or morsel driven parallelism has come up several times, most recently in the Discord channel: https://discord.com/channels/885562378132000778/1166447479609376850/1198721509842235402

I would like to document what is going on so that we can refer people to the documentation when these questions inevitably arise in the future

## What changes are included in this PR?
1. Add a note on pull execution and why Datafusion doesn't implement morsel driven paralleism directly
2. Add a note about using tokio as the thread pool

## Are these changes tested?
Existing CI checks

## Are there any user-facing changes?
More docs

### CODE REVIEW

File: datafusion/core/src/lib.rs [hunk 1]
```diff
@@ -342,16 +342,20 @@
...
+//! Calling [`execute`] produces 1 or more partitions of data,
+//! as a [`SendableRecordBatchStream`], which implements a pull based execution
+//! API. Calling `.next().await` will incrementally compute and return the next
+//! [`RecordBatch`]. Balanced parallelism is achieved using [Volcano style]
+//! "Exchange" operations implemented by  [`RepartitionExec`].
+//!
+//! While some recent research such as [Morsel-Driven Parallelism] describes challenges
+//! with the pull style Volcano execution model on NUMA architectures, in practice DataFusion achieves
+//! similar scalability as systems that use morsel driven approach such as DuckDB.
+//! See the [DataFusion paper submitted SIGMOD] for more details.
```
Comment (tustvold) [inline datafusion/core/src/lib.rs:358]:
```suggestion
//! See the [DataFusion paper submitted to SIGMOD] for more details.
```

File: datafusion/core/src/lib.rs [hunk 2]
```diff
@@ -364,8 +368,26 @@
 //!
 //! [`RepartitionExec`]: https://docs.rs/datafusion/latest/datafusion/physical_plan/repartition/struct.RepartitionExec.html
 //! [Volcano style]: https://w6113.github.io/files/papers/volcanoparallelism-89.pdf
+//! [Morsel-Driven Parallelism]: https://db.in.tum.de/~leis/papers/morsels.pdf
+//! [DataFusion paper submitted SIGMOD]: https://github.com/apache/arrow-datafusion/files/13874720/DataFusion_Query_Engine___SIGMOD_2024.pdf
```
Comment (tustvold) [inline datafusion/core/src/lib.rs:372]:
```suggestion
//! [DataFusion paper submitted to SIGMOD]: https://github.com/apache/arrow-datafusion/files/13874720/DataFusion_Query_Engine___SIGMOD_2024.pdf
```

File: datafusion/core/src/lib.rs [hunk 3]
```diff
@@ -342,16 +342,20 @@
...
 //! [`ScalarValue`] (single constant values) or [`ArrayRef`] (Arrow
 //! Arrays).
 //!
-//! Balanced parallelism is achieved using [`RepartitionExec`], which
-//! implements a [Volcano style] "Exchange".
+//! Calling [`execute`] produces 1 or more partitions of data,
+//! as a [`SendableRecordBatchStream`], which implements a pull based execution
+//! API. Calling `.next().await` will incrementally compute and return the next
+//! [`RecordBatch`]. Balanced parallelism is achieved using [Volcano style]
+//! "Exchange" operations implemented by  [`RepartitionExec`].
```
Comment (tustvold) [inline datafusion/core/src/lib.rs:353]:
```suggestion
//! "Exchange" operations implemented by [`RepartitionExec`].
```

File: datafusion/core/src/lib.rs [hunk 4]
```diff
@@ -364,8 +368,26 @@
...
+//! ## Thread Scheduling
+//!
+//! DataFusion incrementally computes output from a [`SendableRecordBatchStream`]
+//! with `target_partitions` threads. Parallelism is implementing using multiple
+//! [Tokio] [`task`]s, which are executed by threads managed by a tokio Runtime.
+//! While tokio is most commonly used
+//! for asynchronous network I/O, its combination of an efficient, work-stealing
+//! scheduler, first class compiler support for automatic continuation generation,
+//! and exceptional performance makes it a compelling choice for CPU intensive
+//! applications as well, as explained in more detail in [Using Rustlang’s Async Tokio
```
Comment (tustvold) [inline datafusion/core/src/lib.rs:384]:
```suggestion
//! applications as well. This is explained in more detail in [Using Rustlang’s Async Tokio
```

**Review by tustvold — APPROVED** [review]
Just some minor copy suggestions

File: datafusion/core/src/lib.rs [hunk 5]
```diff
@@ -364,8 +368,26 @@
...
+//! [Morsel-Driven Parallelism]: https://db.in.tum.de/~leis/papers/morsels.pdf
+//! [DataFusion paper submitted SIGMOD]: https://github.com/apache/arrow-datafusion/files/13874720/DataFusion_Query_Engine___SIGMOD_2024.pdf
 //! [implementors of `ExecutionPlan`]: https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlan.html#implementors
 //!
+//! ## Thread Scheduling
+//!
+//! DataFusion incrementally computes output from a [`SendableRecordBatchStream`]
+//! with `target_partitions` threads. Parallelism is implementing using multiple
+//! [Tokio] [`task`]s, which are executed by threads managed by a tokio Runtime.
+//! While tokio is most commonly used
```
Comment (alamb) [inline datafusion/core/src/lib.rs:380]: I should probably also point out that this phrasing came from the SIGMOD paper and I think at least @yjshen  and @ozankabak  had a part in authoring it

**Pull request merged**